instance.cc

```diff
@@ -3381,6 +3466,41 @@ float PGInstance::getComputedBiomeHeight(unsigned char b, const vm::vec2 &worldP
     return biomeHeight;
 }
 
+// materials
+void PGInstance::getComputedMaterials(Heightfield &localHeightfield, std::unordered_map<uint8_t, float> &materialsCounts, float &totalMaterialFactors, const vm::vec2 &worldPosition)
```

Comment:
Maps are not very efficient for this purpose (`std::unordered_map<uint8_t, float> &materialsCounts`).

If the range is known then it can be a well-sized `vector` with integer indices and it will run much faster.